walkeeper/src/receive_wal.rs

```diff
@@ -186,37 +182,65 @@ fn request_callback(conf: WalAcceptorConf, timelineid: ZTimelineId, tenantid: ZT
     }
 }
 
-impl ReceiveWalConn {
-    pub fn new(socket: TcpStream, conf: WalAcceptorConf) -> Result<ReceiveWalConn> {
-        let peer_addr = socket.peer_addr()?;
-        let conn = ReceiveWalConn {
-            timeline: None,
-            stream_in: BufReader::new(socket.try_clone()?),
-            stream_out: socket,
+impl<'pg> ReceiveWalConn<'pg> {
+    pub fn new(pg: &'pg mut PostgresBackend) -> Result<ReceiveWalConn<'pg>> {
+        let peer_addr = pg.get_stream_in()?.get_ref().peer_addr()?;
+        Ok(ReceiveWalConn {
+            pg_backend: pg,
             peer_addr,
-            conf,
-        };
-        Ok(conn)
+        })
     }
 
-    fn read_req<T: LeSer>(&mut self) -> Result<T> {
-        // As the trait bound implies, this always encodes little-endian.
-        Ok(T::des_from(&mut self.stream_in)?)
+    // Read and extract the bytes of a `CopyData` message from the postgres instance
+    fn read_msg_bytes(&mut self) -> Result<Bytes> {
+        match self.pg_backend.read_message()? {
+            Some(FeMessage::CopyData(bytes)) => Ok(bytes),
+            Some(msg) => bail!("expected `CopyData` message, found {:?}", msg),
+            None => bail!("connection closed unexpectedly"),
+        }
+    }
+
+    // Read the result of a `CopyData` message sent from the postgres instance
+    //
+    // As the trait bound implies, this always encodes little-endian.
+    fn read_msg<T: LeSer>(&mut self) -> Result<T> {
+        let data = self.read_msg_bytes()?;
+        let mut reader = data.reader();
+        let value = T::des_from(&mut reader)?;
+
+        // We need to check that there's nothing left in the `CopyData` message; that would
+        // indicate something else went wrong.
+        let mut byte = 0_u8;
+        let singleton = std::slice::from_mut(&mut byte);
+        let has_remaining = reader.read(singleton)? != 0;
+        if has_remaining {
+            bail!("unexpected trailing bytes");
+        }
```

Comment:
Ah, I hadn't noticed `has_remaining` -- it looks like that would work. That being said, it does look like using `le_coder` to avoid a reader in the first place would be better. Thanks!